openproblems/tasks/spatial_decomposition/_utils.py

```diff
@@ -1,16 +1,83 @@
-from anndata import AnnData
+from typing import Tuple
+from typing import TypedDict
 
+import anndata as ad
 import numpy as np
 
 
-def obs_means(adata: AnnData, cluster_key: str) -> AnnData:
+class SpScBC(TypedDict):
+    spatial: str
+    single_cell: str
+
+
+def merge_sc_and_sp(
+    adata_sc: ad.AnnData,
+    adata_sp: ad.AnnData,
+    batch_key: str = "modality",
+    batch_categories: SpScBC = dict(spatial="sp", single_cell="sc"),
+) -> ad.AnnData:
+
+    # merge single cell and spatial data
+    adata_merged = adata_sp.concatenate(
+        adata_sc,
+        batch_key=batch_key,
+        batch_categories=(batch_categories["spatial"], batch_categories["single_cell"]),
+        join="outer",
+        index_unique=None,
+    )
+
+    # get attributes from single cell and spatial data
+    sc_obs = set(adata_sc.obs.columns.tolist())
+    sp_obs = set(adata_sp.obs.columns.tolist())
+    sc_obsm = set(list(adata_sc.obsm.keys()))
+    sp_obsm = set(list(adata_sp.obsm.keys()))
+
+    # store unique attributes for cleanup in split
+    attr_dict = dict(
+        sc_obs_only=list(sc_obs.difference(sp_obs)),
+        sp_obs_only=list(sp_obs.difference(sc_obs)),
+        sc_obsm_only=list(sc_obsm.difference(sp_obsm)),
+        sp_obsm_only=list(sp_obsm.difference(sc_obsm)),
+    )
+
+    adata_merged.uns["attr_dict"] = attr_dict
+
+    return adata_merged
+
+
+def split_sc_and_sp(
+    adata_merged: ad.AnnData,
+    batch_key: str = "modality",
+    batch_categories: SpScBC = dict(spatial="sp", single_cell="sc"),
+) -> Tuple[ad.AnnData, ad.AnnData]:
+
+    # split single cell and spatial data
+    is_sp = adata_merged.obs[batch_key] == batch_categories["spatial"]
+    adata_sp = adata_merged[is_sp, :]
+    adata_sc = adata_merged[~is_sp, :]
+
+    # clean objects
```

Comment:
again, unclear why we need to "delete" obs? Is it because a bunch of `np.nan` are added in concatenation? Not sure is huge problem in terms of storage. I'd approach it differentily by removing obs altogether and except for `cluster_label` of `adata_sc` ? but maybe again I'm missing something so feel free to disregard

Reply:
Yes exactly, I thought it was a bit unnecessary to have all these `np.nan` columns, but if you don't mind I'm fine with removing this part - I was honestly a bit hesitant to even put it in there!